fitlins/interfaces/nistats.py

```diff
@@ -196,46 +196,77 @@ def _run_interface(self, runtime):
         # Only keep files which match all entities for contrast
         stat_metadata = _flatten(self.inputs.stat_metadata)
         input_effects = _flatten(self.inputs.effect_maps)
+        input_variances = _flatten(self.inputs.variance_maps)
 
         filtered_effects = []
+        filtered_variances = []
         names = []
-        for m, eff in zip(stat_metadata, input_effects):
+        for m, eff, var in zip(stat_metadata, input_effects, input_variances):
             if _match(out_ents, m):
                 filtered_effects.append(eff)
+                filtered_variances.append(var)
                 names.append(m['contrast'])
 
-        # Dummy code contrast of input effects
-        design_matrix = pd.get_dummies(names)
+        mat = pd.get_dummies(names)
+        contrasts = prepare_contrasts(self.inputs.contrast_info, mat.columns)
 
-        # Fit single model for all inputs
-        model.fit(filtered_effects, design_matrix=design_matrix)
+        # Only fit model if any non-FEMA contrasts at this level
+        if any([c[2] != 'FEMA' for c in contrasts]):
+            if len(filtered_effects) < 2:
+                raise RuntimeError(
+                    "At least two inputs are required for a 't' for 'F' "
+                    "second level contrast")
+            else:
+                model = level2.SecondLevelModel(smoothing_fwhm=smoothing_fwhm)
+                model.fit(filtered_effects, design_matrix=mat)
 
-        for name, weights, contrast_type in prepare_contrasts(
-          self.inputs.contrast_info, design_matrix.columns.to_list()):
+        for name, weights, contrast_type in contrasts:
             contrast_metadata.append(
                 {'contrast': name,
                  'stat': contrast_type,
                  **out_ents})
 
-            maps = model.compute_contrast(
-                second_level_contrast=weights,
-                second_level_stat_type=contrast_type,
-                output_type='all')
+            # Pass-through happens automatically as it can handle 1 input
+            if contrast_type == 'FEMA':
+                # Filter effects and variances based on weights
+                ix = mat.iloc[:, weights[0].astype('bool')].sum(axis=1)
```

Comment:
I played with this, and I'm not sure it does what you want:

```Python
>>> import numpy as np
>>> import pandas as pd
>>> mat = pd.get_dummies(['a', 'b', 'c'])
>>> effects = ['d', 'e', 'f']
>>> ix = mat.iloc[:, np.array([1, 0, 1], dtype=bool)].sum(axis=1)
>>> ix
0    1
1    0
2    1
dtype: int64
>>> np.array(effects)[ix]                                                                      
array(['e', 'd', 'e'], dtype='<U1')
```

I think what you want is:

```Python
>>> np.array(effects)[ix]                                                                      
array(['d', 'f'], dtype='<U1')
```

So:

```suggestion
                ix = mat.iloc[:, weights[0].astype(bool)].sum(axis=1).astype(bool)
```

But isn't `mat` just an identity matrix? So isn't this doubling up?

```suggestion
                ix = weights[0].astype(bool)
```

Not sure if I'm missing something here, though...

Reply:
Yes, what I want is to get the column that are non-zero in a contrast. 

I think i am maybe doubling up let me think about it, but I'm trying to leave the design matrix the same across both types of contrasts.

Reply:
I think your suggestion might work though... For some reason I remember the `sum` being important but maybe not.